stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,304 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_GP_MATERN32_COV_HPP
+#define STAN_MATH_PRIM_MAT_FUN_GP_MATERN32_COV_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/divide_columns.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive_finite.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <stan/math/prim/scal/fun/divide.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/meta/size_of.hpp>
+#include <cmath>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns a Matern 3/2 covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{3}d(x, x')}{l})exp(-\frac{\sqrt{3}d(x, x')}{l})
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type for each scalar
+ * @tparam T_s type of parameter of sigma
+ * @tparam T_l type of parameter length scale
+ *
+ * @param x std::vector of scalars that can be used in squared distance
+ * @param length_scale length scale
+ * @param sigma marginal standard deviation or magnitude
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const T_l &length_scale) {
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = size_of(x);
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  const char *function = "gp_matern32_cov";
+  for (size_t i = 0; i < x_size; ++i)
+    for (size_t ii = i; ii < x_size; ++ii)
+      check_size_match(function, "x row", size_of(x[i]), "x's other row",
+                       size_of(x[ii]));
+
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan(function, "x", x[n]);
+
+  check_positive_finite(function, "magnitude", sigma);
+  check_positive_finite(function, "length scale", length_scale);
+
+  T_s sigma_sq = square(sigma);
+  typename return_type<double, T_l>::type root_3_inv_l
+      = sqrt(3.0) / length_scale;
+  typename return_type<double, T_l>::type neg_root_3_inv_l
+      = -1.0 * sqrt(3.0) / length_scale;
+
+  for (size_t i = 0; i < x_size; ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      typename return_type<T_x, T_s, T_l>::type distance
```

Comment:
This type is incorrect. It should be `T_x`.

`return_type<T_x, T_s, T_l>::type` will be `stan::math::var` if any of the three are `stan::math::var`. You'll be getting overpromotion if `T_x` is `double`. Meaning if `T_x` is actually `double`, then the `sqrt(squared_distance(x[i], x[j)))` will also be `double`, but if `T_s` happens to be `stan::math::var`, we've now constructed an autodiff variable and placed it on the stack. So please change this to `T_x`.

Reply:
[question]
do we not have a `distance` function?